[FLINK-16271][python] Introduce ArrowPythonScalarFunctionOperator for vectorized Python UDF execution

dianfu opened the pull request:
## What is the purpose of the change

*This pull request introduces  ArrowPythonScalarFunctionOperator for vectorized Python UDF execution.*

## Brief change log

  - *Introduces `ArrowPythonScalarFunctionRunner` which is responsible for serializing the input elements into Arrow format and sending them out*
  - *Introduces `ArrowPythonScalarFunctionOperator` is the operator for vectorized Python UDF execution which calling ArrowPythonScalarFunctionRunner to writing the input elements out. It also deserializes the execution results and sending them to the downstream operator*
  - *Introduces configuration `python.fn-execution.arrow.batch.size` which controls the maximum number of elements included in an Arrow batch*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added test cases in PythonConfigTest and PythonOptionsTest which verify that the config `python.fn-execution.arrow.batch.size`*
  - *Added tests ArrowPythonScalarFunctionRunnerTest, ArrowPythonScalarFunctionOperatorTest and BaseRowArrowPythonScalarFunctionOperatorTest*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (Will add doc when end-to-end functionality works)

hequn8128 commented on flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/scalar/arrow/ArrowPythonScalarFunctionOperator.java:
```diff
@@ -0,0 +1,126 @@
+ * Arrow Python {@link ScalarFunction} operator for the old planner.
+ */
+@Internal
+public class ArrowPythonScalarFunctionOperator extends AbstractRowPythonScalarFunctionOperator {
```
It seems it's better to add a base class for the arrow operator, similar to the runner class structure. More code could be reused. For example, most code in `ArrowPythonScalarFunctionOperator` and `BaseRowArrowPythonScalarFunctionOperator` could be reused.

dianfu replied:
I'm afraid that we cannot do that. Currently ArrowPythonScalarFunctionOperator and PythonScalarFunctionOperator extends the same abstract base class, while BaseRowArrowPythonScalarFunctionOperator and BaseRowPythonScalarFunctionOperator share the same abstract base class. This is a little different from the runner because the operators from the same planner could share more code. 
Besides, it seems that there is no much code duplication between ArrowPythonScalarFunctionOperator and BaseRowArrowPythonScalarFunctionOperator and only open/close method could be reused between them(about 20 lines of code). Definitely agree with you that we should remove code duplication as much as possible. However, it's not easy to remove all the code duplication as we want to support two kinds of planners, two kinds of udfs. We can only make the code duplication as minimum as possible. What's your thought?

hequn8128 replied:
ok, let's keep it.

hequn8128 commented on flink-python/src/test/java/org/apache/flink/table/runtime/utils/PassThroughArrowPythonScalarFunctionRunner.java:
```diff
@@ -0,0 +1,87 @@
+/**
+ * An {@link ArrowPythonScalarFunctionRunner} that just return the input elements as the execution results.
+ */
+public class PassThroughArrowPythonScalarFunctionRunner extends ArrowPythonScalarFunctionRunner {
```
I'm wondering if we can create a `CommonPassThroughScalarFunctionRunner<IN>` that can be used for both arrow and non-arrow, both row and baserow, so that we don't need so many duplicated classes.

dianfu replied:
I have tried to do that. However I find that it's impossible as these four runners extends different classes. Do you have any suggestions on how to do that? I'm very glad to improve them if it's possible.

dianfu replied:
Good suggestion. As discussed offline, there are currently two PassThrough runners which are used for arrow and non-arrow cases separately.

dianfu commented on flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/scalar/arrow/BaseRowArrowPythonScalarFunctionOperator.java:
```diff
@@ -0,0 +1,132 @@
+			reader.loadNextBatch();
+			VectorSchemaRoot root = reader.getVectorSchemaRoot();
+			if (arrowReader == null) {
+				arrowReader = ArrowUtils.createBaseRowArrowReader(root);
```
It assumes that the execution results are already stored in the `VectorSchemaRoot` before creating the ArrowReader. So I'm afraid that we cannot move it to open.

dianfu commented:
@hequn8128 Appreciated your review. I have addressed your comments with a few exceptions. Please help to take a look at if it makes sense to you. Thanks a lot.

dianfu commented:
@hequn8128 Thanks for the review. Updated.
